src/main/java/com/google/sps/data/UserQuery.java

```diff
@@ -14,47 +14,128 @@
     this.authors = new ArrayList<String>();
-    this.recipeName = SearchUtils.parseRecipeName(standardizedQuery);
   }
 
   /**
-   * Takes a new tag or ingredient to search for and adds it to the tagsAndIngredients attribute.
-   * TODO: Implement ability to avoid duplicates
+   * Returns a filter which can be used to construct a Datastore query based on the instance variables
+   * of this object. If the object has no data which can be used to construct filters, it will return
+   * a single filter for making sure recipes are public.
    */
-  public void addTag(String tag) {
-    this.tagsAndIngredients.add(tag);
+  public Query.Filter createSearchFilter() {
+    // ArrayList which will hold filters for name, authors, tags, and visibility.
+    ArrayList<Query.Filter> searchFilters = new ArrayList<Query.Filter>();
+    
+    // Creates and adds filters to the list if there is data to make the filter
+    if (tagsAndIngredients.size() > 0){
+      searchFilters.add(createTagsFilter());
+    }
+    if (authors.size() > 0) {
+      searchFilters.add(createAuthorsFilter());
+    }
+    if (recipeName != null && !recipeName.equals("")) {
+      searchFilters.add(new Query.FilterPredicate("name", Query.FilterOperator.EQUAL, recipeName.toLowerCase().trim()));
+    }
+
+    // Adds a filter which makes sure the recipes being searched for are public
+    searchFilters.add(new Query.FilterPredicate("published", Query.FilterOperator.EQUAL, true));
+    
+    if (searchFilters.size() == 1) {
+      return searchFilters.get(0);
+    } else {
+      return new Query.CompositeFilter(
+        Query.CompositeFilterOperator.AND, searchFilters
+      );
+    }
   }
 
-  public ArrayList<String> getTagsAndIngredients() {
-    return this.tagsAndIngredients;
+  /*
+   * Takes a tag a returns a query filter which can be used to check if the tag is in the tag or
+   * ingredient list of the entity property.
+   */
+  private Query.CompositeFilter createSingleTagFilter(String tag) {
+    // Filter for whether the tag is in the ingredient list
+    Query.Filter ingredientFilter = new Query.FilterPredicate("ingredients", Query.FilterOperator.IN, tag);
+      
+    // Filter for whether the tag is in the tag list
+    Query.Filter tagFilter = new Query.FilterPredicate("tags", Query.FilterOperator.IN, tag);
+
+    // Returns a composite filter for whether the tag is in one of the arraylists
+    return new Query.CompositeFilter(
+      Query.CompositeFilterOperator.OR, Arrays.asList(ingredientFilter, tagFilter)
+    );
   }
 
-  public ArrayList<String> getAuthors() {
-    return this.authors;
+  /**
+   * Returns a filter for recipes which have all of the tags and ingredients in the collection
+   * present.
+   */
+  private Query.CompositeFilter createTagsFilter() {
+    // Implementation of collection interface methods to handle future change of ArrayList to set
+    Iterator<String> tagsAndIngredientsIterator = tagsAndIngredients.iterator();
```

Comment:
You can use a for each loop, instead of dealing with the iterator. That's what the for each loop does behind the scenes anyway (https://stackoverflow.com/questions/29601970/what-happens-behind-the-scene-when-i-write-for-loop-like-this)

Reply:
Fixed